content/academy/expert_scraping_with_apify/solutions/saving_stats.md

```diff
@@ -0,0 +1,180 @@
+            offer: element.find('.a-price .a-offscreen').text().trim(),
+        });
+    }
+};
+```
+
+## [](#saving-stats-with-dataset-items) Saving stats with dataset items
+
+Still in the **handleOffers** function, we need to add a few extra keys to the items which are pushed to the dataset. Luckily, all of the data required by the task is easily accessible in the context object.
+
+```JavaScript
+exports.handleOffers = async ({ $, request, crawler: { requestQueue } }, dataset) => {
+    const { data } = request.userData;
+
+    const { asin } = data;
+
+    for (const offer of $('#aod-offer')) {
+        tracker.incrementASIN(asin);
+        // Add 1 to totalSaved for every offer
+        Stats.increment();
+
+        const element = $(offer);
+
+        await dataset.pushData({
+            ...data,
+            sellerName: element.find('div[id*="soldBy"] a[aria-label]').text().trim(),
+            offer: element.find('.a-price .a-offscreen').text().trim(),
+            // Store the handledAt date or current date if that is undefined
+            dateHandled: request.handledAt || new Date().toISOString(),
+            // Access the number of retries on the request object
+            numberOfRetries: request.retryCount,
+            // Grab the number of pending requests from the requestQueue
+            currentPendingRequests: (await requestQueue.getInfo()).pendingRequestCount,
+        });
+    }
+};
+```
+
+## [](#quiz-answers) Quiz answers
+
+**Q: Why might you want to store statistics about an actor's run (or a specific request)?**
+
+**A:** If certain types of requests are error-prone, you might want to save stats about the run to look at them later to either eliminate or better handle the errors. Things like **dateHandled** can be generally useful information.
+
+**Q: In our Amazon scraper, we are trying to store the number of retries of a request once its data is pushed to the dataset. Where would you get this information? Where would you store it?**
+
+**A:** This information is available directly on the request object under the property **retryCount**.
+
+**Q: We are building a new imaginary scraper for a website that sometimes displays captchas at unexpected times, rather than displaying the content we want. How would you keep a count of the total number of captchas hit for the entire run? Where would you store this data? Why?**
+
+**A:** First, build a function that detects if the captcha has been hit. If so, it will throw an error and add to the **numberOfCaptchas** count. This data might be stored on a persisted state object to help better assess the anti-scraping mitigation techniques the scraper should be used.
+
+**Q: Is storing these types of values necessary for every single actor?**
+
+**A:** For small actors, it might be a waste of time to do this. For large-scale actors, it can be extremely helpful when debugging and most definitely worth the extra 10-20 minutes of development time.
```

Comment:
Let's add that the default SDK statistics class might be enough for simple stats. We should probably mention it in the solution as well as the the histogram is very nice